lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -0,0 +1,111 @@
+import Amazonka.Env (Env, Env' (..))
+import Amazonka.HTTP (retryRequest)
+import Amazonka.Lens ((^.))
+import Amazonka.Prelude
+import Amazonka.SSO.GetRoleCredentials as SSO
+import qualified Amazonka.SSO.Types as SSO (RoleCredentials (..))
+import Amazonka.Types
+import qualified Control.Exception as Exception
+import Control.Exception.Lens (handling_, _IOException)
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.Aeson (FromJSON, decodeFileStrict)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import qualified Network.HTTP.Client as Client
+
+data Cache = Cache
+  { accessToken :: Text,
+    region :: Region
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON)
+
+fromSSO ::
+  (MonadIO m, Foldable withAuth) =>
+  FilePath ->
+  Region ->
+  Text ->
+  Text ->
+  Env' withAuth ->
+  m Env
+fromSSO cache ssoRegion accountId roleName env = do
+  auth <- liftIO $ fetchAuthInBackground getCredentials
+  pure $ env {envAuth = Identity auth}
+  where
+    getCredentials = do
+      Cache {..} <- readCache cache
+
+      let ssoEnv = env {envRegion = ssoRegion}
+
+      resp <-
+        runResourceT $
+          sendWithoutAuth ssoEnv $
+            SSO.newGetRoleCredentials roleName accountId accessToken
+
+      let mCreds = do
+            rc <- resp ^. SSO.getRoleCredentialsResponse_roleCredentials
+            roleCredentialsToAuthEnv rc
+
+      case mCreds of
+        Nothing -> fail "sso:GetRoleWithCredentials returned no credentials."
+        Just c -> pure c
+
+sendWithoutAuth ::
+  ( MonadResource m,
+    AWSRequest a,
+    Foldable withAuth
+  ) =>
+  Env' withAuth ->
+  a ->
+  m (AWSResponse a)
+sendWithoutAuth env req = do
+  eResponse <- retryRequest env req
+  either (liftIO . Exception.throwIO) (pure . Client.responseBody) eResponse
+
+relativeCacheFile :: MonadIO m => Text -> m FilePath
+relativeCacheFile startUrl = do
+  let sha1 = show $ Crypto.hashSHA1 $ Text.encodeUtf8 startUrl
+  pure $ "/.aws/sso/cache/" <> sha1 <> ".json"
+
+readCache :: MonadIO m => FilePath -> m Cache
+readCache p = liftIO $
+  handling_ _IOException err $ do
+    mCache <- decodeFileStrict p
+    maybe err pure mCache
+  where
+    err =
+      Exception.throwIO $
+        InvalidFileError $
+          "Unable to read SSO cache. "
+            <> Text.pack p
+            <> " is missing or invalid."
+
+roleCredentialsToAuthEnv :: SSO.RoleCredentials -> Maybe AuthEnv
+roleCredentialsToAuthEnv rc =
+  AuthEnv
+    <$> (AccessKey . Text.encodeUtf8 <$> SSO.accessKeyId rc)
+    <*> (overSensitive (SecretKey . Text.encodeUtf8) <$> SSO.secretAccessKey rc)
+    <*> pure (overSensitive (SessionToken . Text.encodeUtf8) <$> SSO.sessionToken rc)
+    <*> pure (expirationToExpires <$> SSO.expiration rc)
+
+expirationToExpires :: Integer -> ISO8601
+expirationToExpires = Time . posixSecondsToUTCTime . fromInteger
```

Comment:
Can you please hang this under a `where` clause in `roleCredentialsToAuthEnv`?

Reply:
Sure, care to elaborate why?

At $work, we have a convention to always make something that _can_ be top-level top-level -- we find it helps with thinking about parts in isolation (vs lumping the complexity of this function into the other one as a single "thing"), results in more type signatures (being now non-optional), and will very often end up being re-used (my need to move `configFileRelative` to accomplish this PR muddies its diff when it wouldn't need to had this been followed).

Not pushing back at all, just curious to hear the other side.